### DISCUSSION TITLE
Logic Behind Client Information Sync

### DISCUSSION THREAD

**Discussion by RohanDoshi21** [wind-c/comqtt#79] [Q&A] [answered]
While reading other discussions and a code ( Mochi-broker modified code ),

From my understanding,
1. Client information is being synced across all the nodes using raft for every `processSubscribe` function using a custom hook ( `OnSubscribed` ) defined in `cluster\events.go`

2. You are storing the `node-id` on which a particular client is connected.
    So each node knows on which node a particular client is connected.

3. Now for every publish Message, you are filtering out clients, connected to a node using `sharedFilters`.
    And then using grpc to send out the same packet to that particular node. ( Packet forwarding to the node where the client is connected ).  Now this node will again process the packet and send it out to its intended connected subscribers.
    
Is my understanding correct @wind-c? Also, I further want to know what the logic behind `sharedFilters` is and how they are used

**Comment by wind-c** [comment, marked as answer]
Each node saves the subscription tree of other nodes (only the subscription not the client, which reduces memory footprint) and generates the corresponding routing table (map[subscription]nodes), subscription data is synchronized between nodes via raft. The logical flow is publish -> local subscription tree -> local subscription tree of the other nodes -> routing table -> grpc to node. The same logic applies to shared subscriptions. The difference is that the shared subscription takes precedence on the local node, and if the local node does not have this shared subscription, it will randomly select the node with the shared subscription to send the message. A non-shared subscription is sent to all nodes that have the subscription.

Have I explained myself clearly? If you have any questions please leave a message!

**Reply by RohanDoshi21** [reply]
Thanks, got it!
